### PR TITLE
TITAN-1367: Use correlation IDs to keep responses separate for different requests

### PR DESCRIPTION
**Pull request by peejaybee**

### CODE REVIEW

**Comment by peejaybee** [comment]
This is the ultimate solution to #14 
The external API is identical to that in major version 1, but since v1 does not send back the correlation ID, v2 clients are not compatible with v1 servers.

File: lib/bunny_burrow/client.rb [hunk 1]
```diff
@@ -1,44 +1,63 @@
...
 module BunnyBurrow
   class Client < Base
-    def publish(payload, routing_key)
-      result = nil
 
-      # when creating a queue, a blank name indicates we want the AMPQ broker
-      # to generate a unique name for us. Also note that this queue will be on
-      # the default exchange
-      reply_to = channel.queue('', exclusive: true, auto_delete: true)
+    attr_accessor :correlation_id, :details, :result
```
Comment (johann-koebbe-fidelity) [inline lib/bunny_burrow/client.rb]: What's the purpose of turning details and result into instance variables?

File: lib/bunny_burrow/server.rb [hunk 2]
```diff
@@ -28,6 +28,12 @@ def subscribe(routing_key, queue_name = routing_key, &block)
             properties: properties
           }
 
+          reply_options = {
+              routing_key: properties.reply_to,
```
Comment (johann-koebbe-fidelity) [inline lib/bunny_burrow/server.rb]: Is this indentation correct?

File: lib/bunny_burrow/client.rb [hunk 3]
```diff
@@ -1,44 +1,63 @@
...
+        persistence: false,
+        correlation_id: correlation_id
       }
 
       topic_exchange.publish(payload.to_json, options)
 
-      Timeout.timeout(timeout) do
-        reply_to.subscribe do |_, _, payload|
-          details[:response] = payload if log_response?
-          log "Receiving #{details}"
```
Comment (johann-koebbe-fidelity) [inline lib/bunny_burrow/client.rb]: So with these changes, we're losing the logging of the receipt of the response?

File: lib/bunny_burrow/server.rb (hunk 2 above)
Comment (peejaybee) [inline lib/bunny_burrow/server.rb]: No, it is not; fixed.

File: lib/bunny_burrow/client.rb [hunk 4]
```diff
@@ -1,44 +1,63 @@
...
+    attr_accessor :correlation_id, :details, :result
+
+    def initialize()
+      super
+
+      subscribe_to_replies
+    end
+
+    def subscribe_to_replies
+      Timeout.timeout(timeout) do
```
Comment (stlcardsfan) [inline lib/bunny_burrow/client.rb]: What is this timeout for?

Comment (peejaybee) [inline lib/bunny_burrow/client.rb]: It's so we get a timeout if we can't subscribe.

**Comment by peejaybee** [comment]
I've decided to use direct reply-to (https://www.rabbitmq.com/direct-reply-to.html) instead of a declared reply-to queue and correlation IDs.  Closing/

**Pull request closed without merging**